src/neuroglancer/datasource/graphene/base.ts

```diff
@@ -0,0 +1,111 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {mat4} from 'neuroglancer/util/geom';
+import {ShardingParameters} from 'neuroglancer/datasource/precomputed/base';
+
+export const PYCG_APP_VERSION = 1;
+
+export enum VolumeChunkEncoding {
+  RAW,
+  JPEG,
+  COMPRESSED_SEGMENTATION
+}
+
+export class VolumeChunkSourceParameters {
+  url: string;
+  encoding: VolumeChunkEncoding;
+  sharding: ShardingParameters|undefined;
+
+  static RPC_ID = 'graphene/VolumeChunkSource';
+}
+
+
+export class ChunkedGraphSourceParameters {
+  url: string;
+
+  static RPC_ID = 'graphene/ChunkedGraphSource';
+}
+
+export class MeshSourceParameters {
+  manifestUrl: string;
+  fragmentUrl: string;
+  lod: number;
+  sharding: Array<ShardingParameters>|undefined;
+
+  static RPC_ID = 'graphene/MeshSource';
+}
+
+export class MultiscaleMeshMetadata {
+  transform: mat4;
+  lodScaleMultiplier: number;
+  vertexQuantizationBits: number;
+  sharding: Array<ShardingParameters>|undefined;
+}
+
+/*
+temporary solution to deal with cors error if the middle auth credential is passed to fetch request to GCS.
+because the authorization header is not in Access-Control-Allow-Headers
+*/
+
+import {cancellableFetchSpecialOk as cancellableFetchSpecialOkOrig, SpecialProtocolCredentialsProvider} from 'neuroglancer/util/special_protocol_request';
+import { CancellationToken, uncancelableToken } from 'neuroglancer/util/cancellation';
+import { ResponseTransform } from 'neuroglancer/util/http_request';
+
+const GCS_ORIGIN = 'https://storage.googleapis.com';
+
+export async function cancellableFetchSpecialOk<T>(
+  credentialsProvider: SpecialProtocolCredentialsProvider, url: string, init: RequestInit,
+  transformResponse: ResponseTransform<T>,
+  cancellationToken: CancellationToken = uncancelableToken): Promise<T> {
+    if ((new URL(url)).origin === GCS_ORIGIN) {
+      credentialsProvider = undefined;
+    }
+
+    return cancellableFetchSpecialOkOrig(credentialsProvider, url, init, transformResponse, cancellationToken);
+}
+
+import {fetchSpecialHttpByteRange as fetchSpecialHttpByteRangeOrig} from 'neuroglancer/util/byte_range_http_requests';
+import { Uint64 } from 'neuroglancer/util/uint64';
+
+export function fetchSpecialHttpByteRange(
+  credentialsProvider: SpecialProtocolCredentialsProvider, url: string,
+  startOffset: Uint64|number, endOffset: Uint64|number,
+  cancellationToken: CancellationToken): Promise<ArrayBuffer> {
+    if ((new URL(url)).origin === GCS_ORIGIN) {
```

Comment:
This seems kind of risky in terms of avoiding leaking the credentials --- e.g. if Neuroglancer made GCS requests slightly differently.  If I understand correctly, the intention is that you have one credentials provider for calls to the chunk graph and mesh services, and then don't use it for fetching the static data from GCS.  I think it would be better for the callers to track explicitly which credentials provider they should use, since in the future you might not use GCS for the static data, and you might want to supply separate credentials to the static data service.

Reply:
I just removed these functions, they were only intended to be temporary, if I still need similar conditional logic on when to pass credentials, I'll handle it in the calling function.